### DISCUSSION TITLE
Does this work with Arduino Boards?

### DISCUSSION THREAD

**Discussion by polyphonic13** [hsaturn/TinyMqtt#91] [Q&A] [answered]
I am attempting to create an MQTT broker on an Arduino MKR Wifi 1010. I found this via the Arduino page here https://www.arduino.cc/reference/en/libraries/tinymqtt/, which states:

>  you should be able to use it on all the Arduino boards.

The repo though says:

> TinyMqtt is a small, fast and capable Mqtt Broker and Client for Esp8266 / Esp32 / Esp WROOM

Is there support for standard Arduino boards, or is this only available for Esp boards? 
When attempting to use any of the example files from the repo, the Arduino IDE throws complaints on all of them, e.g. the following error  from this example https://github.com/hsaturn/TinyMqtt/blob/main/examples/tinymqtt-test/tinymqtt-test.ino

`FQBN: arduino:samd:mkrwifi1010
Using board 'mkrwifi1010' from platform in folder: /Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13
Using core 'arduino' from platform in folder: /Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13

Detecting libraries used...
/Users/paulhoover/Library/Arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4/bin/arm-none-eabi-g++ -mcpu=cortex-m0plus -mthumb -c -g -Os -w -std=gnu++11 -ffunction-sections -fdata-sections -fno-threadsafe-statics -nostdlib --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -w -x c++ -E -CC -DF_CPU=48000000L -DARDUINO=10607 -DARDUINO_SAMD_MKRWIFI1010 -DARDUINO_ARCH_SAMD -DUSE_ARDUINO_MKR_PIN_LAYOUT -D__SAMD21G18A__ -DUSB_VID=0x2341 -DUSB_PID=0x8054 -DUSBCON "-DUSB_MANUFACTURER=\"Arduino LLC\"" "-DUSB_PRODUCT=\"Arduino MKR WiFi 1010\"" -DUSE_BQ24195L_PMIC -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS/4.5.0/CMSIS/Include/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0/CMSIS/Device/ATMEL/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated-avr-comp -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/variants/mkrwifi1010 /private/var/folders/vc/141n93x52v31p772rpyz286r0000gn/T/arduino/sketches/9663BCB8FC38B4CFC6AF4F7FF317BD2E/sketch/plant-watering04-tiny-mqtt.ino.cpp -o /dev/null
Alternatives for TinyConsole.h: [TinyConsole@0.4.5]
ResolveLibrary(TinyConsole.h)
  -> candidates: [TinyConsole@0.4.5]
/Users/paulhoover/Library/Arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4/bin/arm-none-eabi-g++ -mcpu=cortex-m0plus -mthumb -c -g -Os -w -std=gnu++11 -ffunction-sections -fdata-sections -fno-threadsafe-statics -nostdlib --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -w -x c++ -E -CC -DF_CPU=48000000L -DARDUINO=10607 -DARDUINO_SAMD_MKRWIFI1010 -DARDUINO_ARCH_SAMD -DUSE_ARDUINO_MKR_PIN_LAYOUT -D__SAMD21G18A__ -DUSB_VID=0x2341 -DUSB_PID=0x8054 -DUSBCON "-DUSB_MANUFACTURER=\"Arduino LLC\"" "-DUSB_PRODUCT=\"Arduino MKR WiFi 1010\"" -DUSE_BQ24195L_PMIC -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS/4.5.0/CMSIS/Include/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0/CMSIS/Device/ATMEL/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated-avr-comp -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/variants/mkrwifi1010 -I/Users/paulhoover/Documents/Arduino/libraries/TinyConsole/src /private/var/folders/vc/141n93x52v31p772rpyz286r0000gn/T/arduino/sketches/9663BCB8FC38B4CFC6AF4F7FF317BD2E/sketch/plant-watering04-tiny-mqtt.ino.cpp -o /dev/null
Alternatives for TinyMqtt.h: [TinyMqtt@0.9.16]
ResolveLibrary(TinyMqtt.h)
  -> candidates: [TinyMqtt@0.9.16]
/Users/paulhoover/Library/Arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4/bin/arm-none-eabi-g++ -mcpu=cortex-m0plus -mthumb -c -g -Os -w -std=gnu++11 -ffunction-sections -fdata-sections -fno-threadsafe-statics -nostdlib --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -w -x c++ -E -CC -DF_CPU=48000000L -DARDUINO=10607 -DARDUINO_SAMD_MKRWIFI1010 -DARDUINO_ARCH_SAMD -DUSE_ARDUINO_MKR_PIN_LAYOUT -D__SAMD21G18A__ -DUSB_VID=0x2341 -DUSB_PID=0x8054 -DUSBCON "-DUSB_MANUFACTURER=\"Arduino LLC\"" "-DUSB_PRODUCT=\"Arduino MKR WiFi 1010\"" -DUSE_BQ24195L_PMIC -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS/4.5.0/CMSIS/Include/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0/CMSIS/Device/ATMEL/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated-avr-comp -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/variants/mkrwifi1010 -I/Users/paulhoover/Documents/Arduino/libraries/TinyConsole/src -I/Users/paulhoover/Documents/Arduino/libraries/TinyMqtt/src /private/var/folders/vc/141n93x52v31p772rpyz286r0000gn/T/arduino/sketches/9663BCB8FC38B4CFC6AF4F7FF317BD2E/sketch/plant-watering04-tiny-mqtt.ino.cpp -o /dev/null
Error while detecting libraries included by /private/var/folders/vc/141n93x52v31p772rpyz286r0000gn/T/arduino/sketches/9663BCB8FC38B4CFC6AF4F7FF317BD2E/sketch/plant-watering04-tiny-mqtt.ino.cpp
/Users/paulhoover/Library/Arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4/bin/arm-none-eabi-g++ -mcpu=cortex-m0plus -mthumb -c -g -Os -w -std=gnu++11 -ffunction-sections -fdata-sections -fno-threadsafe-statics -nostdlib --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -w -x c++ -E -CC -DF_CPU=48000000L -DARDUINO=10607 -DARDUINO_SAMD_MKRWIFI1010 -DARDUINO_ARCH_SAMD -DUSE_ARDUINO_MKR_PIN_LAYOUT -D__SAMD21G18A__ -DUSB_VID=0x2341 -DUSB_PID=0x8054 -DUSBCON "-DUSB_MANUFACTURER=\"Arduino LLC\"" "-DUSB_PRODUCT=\"Arduino MKR WiFi 1010\"" -DUSE_BQ24195L_PMIC -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS/4.5.0/CMSIS/Include/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0/CMSIS/Device/ATMEL/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated-avr-comp -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/variants/mkrwifi1010 -I/Users/paulhoover/Documents/Arduino/libraries/TinyConsole/src -I/Users/paulhoover/Documents/Arduino/libraries/TinyMqtt/src /Users/paulhoover/Documents/Arduino/libraries/TinyConsole/src/TinyConsole.cpp -o /dev/null
/Users/paulhoover/Library/Arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4/bin/arm-none-eabi-g++ -mcpu=cortex-m0plus -mthumb -c -g -Os -w -std=gnu++11 -ffunction-sections -fdata-sections -fno-threadsafe-statics -nostdlib --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -w -x c++ -E -CC -DF_CPU=48000000L -DARDUINO=10607 -DARDUINO_SAMD_MKRWIFI1010 -DARDUINO_ARCH_SAMD -DUSE_ARDUINO_MKR_PIN_LAYOUT -D__SAMD21G18A__ -DUSB_VID=0x2341 -DUSB_PID=0x8054 -DUSBCON "-DUSB_MANUFACTURER=\"Arduino LLC\"" "-DUSB_PRODUCT=\"Arduino MKR WiFi 1010\"" -DUSE_BQ24195L_PMIC -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS/4.5.0/CMSIS/Include/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0/CMSIS/Device/ATMEL/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated-avr-comp -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/variants/mkrwifi1010 -I/Users/paulhoover/Documents/Arduino/libraries/TinyConsole/src -I/Users/paulhoover/Documents/Arduino/libraries/TinyMqtt/src /Users/paulhoover/Documents/Arduino/libraries/TinyConsole/src/TinyString.cpp -o /dev/null
/Users/paulhoover/Library/Arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4/bin/arm-none-eabi-g++ -mcpu=cortex-m0plus -mthumb -c -g -Os -w -std=gnu++11 -ffunction-sections -fdata-sections -fno-threadsafe-statics -nostdlib --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -w -x c++ -E -CC -DF_CPU=48000000L -DARDUINO=10607 -DARDUINO_SAMD_MKRWIFI1010 -DARDUINO_ARCH_SAMD -DUSE_ARDUINO_MKR_PIN_LAYOUT -D__SAMD21G18A__ -DUSB_VID=0x2341 -DUSB_PID=0x8054 -DUSBCON "-DUSB_MANUFACTURER=\"Arduino LLC\"" "-DUSB_PRODUCT=\"Arduino MKR WiFi 1010\"" -DUSE_BQ24195L_PMIC -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS/4.5.0/CMSIS/Include/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0/CMSIS/Device/ATMEL/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated-avr-comp -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/variants/mkrwifi1010 -I/Users/paulhoover/Documents/Arduino/libraries/TinyConsole/src -I/Users/paulhoover/Documents/Arduino/libraries/TinyMqtt/src /Users/paulhoover/Documents/Arduino/libraries/TinyMqtt/src/StringIndexer.cpp -o /dev/null
/Users/paulhoover/Library/Arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4/bin/arm-none-eabi-g++ -mcpu=cortex-m0plus -mthumb -c -g -Os -w -std=gnu++11 -ffunction-sections -fdata-sections -fno-threadsafe-statics -nostdlib --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -w -x c++ -E -CC -DF_CPU=48000000L -DARDUINO=10607 -DARDUINO_SAMD_MKRWIFI1010 -DARDUINO_ARCH_SAMD -DUSE_ARDUINO_MKR_PIN_LAYOUT -D__SAMD21G18A__ -DUSB_VID=0x2341 -DUSB_PID=0x8054 -DUSBCON "-DUSB_MANUFACTURER=\"Arduino LLC\"" "-DUSB_PRODUCT=\"Arduino MKR WiFi 1010\"" -DUSE_BQ24195L_PMIC -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS/4.5.0/CMSIS/Include/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0/CMSIS/Device/ATMEL/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated-avr-comp -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/variants/mkrwifi1010 -I/Users/paulhoover/Documents/Arduino/libraries/TinyConsole/src -I/Users/paulhoover/Documents/Arduino/libraries/TinyMqtt/src /Users/paulhoover/Documents/Arduino/libraries/TinyMqtt/src/TinyMqtt.cpp -o /dev/null
Generating function prototypes...
/Users/paulhoover/Library/Arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4/bin/arm-none-eabi-g++ -mcpu=cortex-m0plus -mthumb -c -g -Os -w -std=gnu++11 -ffunction-sections -fdata-sections -fno-threadsafe-statics -nostdlib --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -w -x c++ -E -CC -DF_CPU=48000000L -DARDUINO=10607 -DARDUINO_SAMD_MKRWIFI1010 -DARDUINO_ARCH_SAMD -DUSE_ARDUINO_MKR_PIN_LAYOUT -D__SAMD21G18A__ -DUSB_VID=0x2341 -DUSB_PID=0x8054 -DUSBCON "-DUSB_MANUFACTURER=\"Arduino LLC\"" "-DUSB_PRODUCT=\"Arduino MKR WiFi 1010\"" -DUSE_BQ24195L_PMIC -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS/4.5.0/CMSIS/Include/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0/CMSIS/Device/ATMEL/ -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino/api/deprecated-avr-comp -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/cores/arduino -I/Users/paulhoover/Library/Arduino15/packages/arduino/hardware/samd/1.8.13/variants/mkrwifi1010 -I/Users/paulhoover/Documents/Arduino/libraries/TinyConsole/src -I/Users/paulhoover/Documents/Arduino/libraries/TinyMqtt/src /private/var/folders/vc/141n93x52v31p772rpyz286r0000gn/T/arduino/sketches/9663BCB8FC38B4CFC6AF4F7FF317BD2E/sketch/plant-watering04-tiny-mqtt.ino.cpp -o /private/var/folders/vc/141n93x52v31p772rpyz286r0000gn/T/arduino/sketches/9663BCB8FC38B4CFC6AF4F7FF317BD2E/preproc/ctags_target_for_gcc_minus_e.cpp
/Users/paulhoover/Documents/code/robots/robot01/packages/plants/plant-watering04-tiny-mqtt/plant-watering04-tiny-mqtt/plant-watering04-tiny-mqtt.ino:11:2: error: #error Unsupported platform
 #error Unsupported platform
  ^~~~~

Using library TinyConsole at version 0.4.5 in folder: /Users/paulhoover/Documents/Arduino/libraries/TinyConsole 
Using library TinyMqtt at version 0.9.16 in folder: /Users/paulhoover/Documents/Arduino/libraries/TinyMqtt 
exit status 1

Compilation error: #error Unsupported platform`

**Comment by hsaturn** [comment, marked as answer]
Hello, this should work for Arduino too, assuming you have an ehternet board.

The error comes from tiny-mqtt-test.ino
``` c++
11 #error Unsupported platform
```

You have to replace here the #error by an include that supports mdns and wificlient.
``` c++
#include <WiFi.h>
#include <ESPmDNS.h>
```

If such libs exists, and are compatible with esp ones, you win !
Else, either you write proxies for Wifi/WifiClient and MDNS classes.

Btw, MDNS is not a dependency of TinyMqtt, ony WiFiClient is needed.

Best regards

**Reply by polyphonic13** [reply]
OK. Thank you! I will give this a try.
